pkg/phlaredb/phlaredb.go

```diff
@@ -146,47 +157,34 @@ func (f *PhlareDB) loop() {
 		select {
 		case <-f.stopCh:
 			return
-		case <-f.headFlushCh():
-			if err := f.Flush(ctx); err != nil {
-				level.Error(f.logger).Log("msg", "flushing head block failed", "err", err)
-				continue
-			}
-			f.runBlockQuerierSync(ctx)
 		case <-blockScanTicker.C:
 			f.runBlockQuerierSync(ctx)
+		case <-headSizeCheck.C:
+			if f.headSize() > maxBlockBytes {
+				f.flushHead(ctx, flushReasonMaxBlockBytes)
+			}
+		case <-f.forceFlush.C:
+			f.flushHead(ctx, flushReasonMaxDuration)
 		case e := <-f.evictCh:
 			f.evictBlock(e)
-		case <-f.headInit:
-			// headFlushCh() may actually be stopCh. When a new head is
-			// initialized, we re-build the select channel list, and get
-			// a valid flush channel of the new head.
-			f.headLock.Lock()
-			f.headInit = make(chan struct{})
-			f.headLock.Unlock()
 		}
 	}
 }
 
-// initHead initializes a new head and signals to the main closing the
-// headInit channel: must only be called with headLock held for writes.
-func (f *PhlareDB) initHead() (err error) {
-	if f.head, err = NewHead(f.phlarectx, f.cfg, f.limiter); err != nil {
-		return err
+func (f *PhlareDB) maxBlockDuration() time.Duration {
+	maxBlockDuration := 5 * time.Second
+	if f.cfg.MaxBlockDuration > maxBlockDuration {
+		maxBlockDuration = f.cfg.MaxBlockDuration
 	}
-	close(f.headInit) // Now can select from f.head.flushCh (headFlushCh).
-	return nil
+	return maxBlockDuration
 }
 
-func (f *PhlareDB) headFlushCh() chan struct{} {
-	f.headLock.RLock()
-	defer f.headLock.RUnlock()
-	if h := f.head; h != nil {
-		return h.flushCh
+func (f *PhlareDB) maxBlockBytes() uint64 {
+	maxBlockBytes := defaultParquetConfig.MaxBlockBytes
+	if f.cfg.Parquet != nil && f.cfg.Parquet.MaxBlockBytes > maxBlockBytes {
```

Comment:
Not too sure how much sense it makes to limit the block size to the default. Maybe someone wanting to run really small clusters might benefit from setting it lower. (Currently we don't expose the limit in CLI/config at all, so this is a theoretical question)

Reply:
Agreed, it makes no sense, overlooked that. Thanks for pointing out!